### PR TITLE
[Keras-Sprint] Modification of the UNet example

### PR DESCRIPTION
**Pull request by ariG23498**
This PR is an effort to modify the Unet example with KerasCV.

[Colab Notebook](https://colab.research.google.com/drive/1JfvohpOkpp3nW0ilxOLgQJqjFDQJ75yl?usp=sharing)

CC: https://github.com/martin-gorner, https://github.com/margaretmz

### CODE REVIEW

File: examples/vision/oxford_pets_image_segmentation.py [hunk 1]
```diff
@@ -1,113 +1,218 @@
...
+So it's important to have a solid understanding of its architecture.
 """
 
-"""shell
-curl -O https://thor.robots.ox.ac.uk/~vgg/data/pets/images.tar.gz
-curl -O https://thor.robots.ox.ac.uk/~vgg/data/pets/annotations.tar.gz
-tar -xf images.tar.gz
-tar -xf annotations.tar.gz
 """
+## Setup and Imports
```
Comment (martin-gorner) [inline examples/vision/oxford_pets_image_segmentation.py:24]: please add "!pip install keras-cv" at the beginning.
I looked at existing KerasCV guides and they all do the installation differently, or don't do it at all. Let's standardize on the simplest.

File: examples/vision/oxford_pets_image_segmentation.py [hunk 2]
```diff
@@ -1,113 +1,218 @@
...
-import os
+Please feel free to tweak the configurations yourself and note how the model training
+changes. This is an excellent exercise to get more understanding of the training
+pipeline.
+"""
 
-input_dir = "images/"
-target_dir = "annotations/trimaps/"
-img_size = (160, 160)
+# Image Config
```
Comment (martin-gorner) [inline examples/vision/oxford_pets_image_segmentation.py:46]: In Python, it is customary to use UPPER CASE for constants.

File: examples/vision/oxford_pets_image_segmentation.py [hunk 3]
```diff
@@ -1,113 +1,218 @@
...
+max_learning_rate = 5e-4
+warmup_epoch_percentage = 0.15
+AUTOTUNE = tf.data.AUTOTUNE
 
-for input_path, target_path in zip(input_img_paths[:10], target_img_paths[:10]):
-    print(input_path, "|", target_path)
+"""
+## Utility Functions
 
+The `unpackage_inputs` is a utility function that is later used to reformat the input
```
Comment (martin-gorner) [inline examples/vision/oxford_pets_image_segmentation.py:65]: This needs a bit more explaining:
- about the KerasCV dictionary format understood by preprocessing layers
- about the fact that the model will not understand that format and will expect a more traditional (image, target) pair
See where this explanation fits best in the text. You can move the "unpackage_inpuuts" function definition there.

File: examples/vision/oxford_pets_image_segmentation.py [hunk 4]
```diff
@@ -1,113 +1,218 @@
...
+import random
+import numpy as np
+from matplotlib import pyplot as plt
+
 """
+## Configuration
 
-import os
+Please feel free to tweak the configurations yourself and note how the model training
+changes. This is an excellent exercise to get more understanding of the training
```
Comment (martin-gorner) [inline examples/vision/oxford_pets_image_segmentation.py:42]: "get more understanding" => "get a better understanding"

File: examples/vision/oxford_pets_image_segmentation.py [hunk 5]
```diff
@@ -1,113 +1,218 @@
...
+    y_true=segmentation_masks,
+    y_pred=None,
+    scale=4,
+    rows=2,
+    cols=2,
+)
+
+"""
+## Model architecture
+The U-Net consists of an encoder for downsammpling and a decoder for upsampling with skip
```
Comment (martin-gorner) [inline examples/vision/oxford_pets_image_segmentation.py:205]: "downsammpling" => "downsampling"

File: examples/vision/oxford_pets_image_segmentation.py [hunk 6]
```diff
@@ -1,113 +1,218 @@
...
+"""
+## Data Augmentation
+We resize both the images and masks to the width/height as specified. Then use KerasCV's
+`RandomFlip`, `RandomRotation` and `RandAugment` to apply image augmentation of random
+flip, random rotation and RandAugment to the train dataset. Here is [a tutorial with more
+details on RandAugment](https://keras.io/examples/vision/randaugment/).
 
-    def __len__(self):
-        return len(self.target_img_paths) // self.batch_size
+We apply only resizing operation to the validation dataset.
```
Comment (martin-gorner) [inline examples/vision/oxford_pets_image_segmentation.py:140]: "We apply only resizing operation to the validation dataset" => "We only apply the resizing operation to the validation dataset"

File: examples/vision/oxford_pets_image_segmentation.py [hunk 7]
```diff
@@ -116,129 +221,281 @@ def get_model(img_size, num_classes):
...
-random.Random(1337).shuffle(target_img_paths)
-train_input_img_paths = input_img_paths[:-val_samples]
-train_target_img_paths = target_img_paths[:-val_samples]
-val_input_img_paths = input_img_paths[-val_samples:]
-val_target_img_paths = target_img_paths[-val_samples:]
-
-# Instantiate data Sequences for each split
-train_gen = OxfordPets(
-    batch_size, img_size, train_input_img_paths, train_target_img_paths
+class WarmUpCosine(
```
Comment (martin-gorner) [inline examples/vision/oxford_pets_image_segmentation.py:293]: Did this help ? I still see a large "hump" in the validation loss.
If this LR schedule does not produce a marked improvement, I'd suggest not to use it.
The LR Schedule definition below is 80+ lines of code which is a lot and takes the focus away from U-NET and Keras-CV.
If the LR schedule really is useful and you want to keep it, I'd recommend going for a simpler tf.keras.optimizers.schedules.PiecewiseConstantDecay with 4 or 5 values for the LR. It will be 2 lines of code.

File: examples/vision/oxford_pets_image_segmentation.py [hunk 8]
```diff
@@ -116,129 +221,281 @@ def get_model(img_size, num_classes):
...
 
-# Configure the model for training.
-# We use the "sparse" version of categorical_crossentropy
-# because our target data is integers.
-model.compile(optimizer="rmsprop", loss="sparse_categorical_crossentropy")
 
-callbacks = [
-    keras.callbacks.ModelCheckpoint("oxford_segmentation.h5", save_best_only=True)
-]
+class DisplayCallback(keras.callbacks.Callback):
```
Comment (martin-gorner) [inline examples/vision/oxford_pets_image_segmentation.py:397]: I like it how you can see the model training! this is very nice.

File: examples/vision/oxford_pets_image_segmentation.py [hunk 9]
```diff
@@ -116,129 +221,281 @@ def get_model(img_size, num_classes):
...
+plt.plot(acc, label="Training accuracy")
+plt.plot(val_acc, label="Validation accuracy")
+plt.title("Training and Validation Accuracy")
+plt.xlabel("Epoch")
+plt.ylabel("Accuracy Value")
+plt.legend()
+plt.show()
+
+"""
+## Prediction with trained model
```
Comment (martin-gorner) [inline examples/vision/oxford_pets_image_segmentation.py:484]: Nice predictions !

File: examples/vision/oxford_pets_image_segmentation.py (hunk 7 above)
Comment (ariG23498) [reply examples/vision/oxford_pets_image_segmentation.py]: We have decided to remove the schedule and also reove the plots from the tutorial.

**Comment by martin-gorner** [comment]
Changes look good. Can you also post the latest Colab for review ? The one linked above does not seem to correspond to the PR (for ex. it has the Cosine decay LR still in it but commented out, which i do not see in the commit. Also the Colab uses PiecewiseConstantDecay which i do not see in the commit either.).

**Review by ianstenbit — COMMENTED** [review]
Thanks for the PR! This example is awesome -- thanks for your contributions to KerasCV to make this UX so nice 😄

File: examples/vision/oxford_pets_image_segmentation.py [hunk 10]
```diff
@@ -1,113 +1,221 @@
...
+the image pixel intensities from `[0, 255]` to the range `[0.0, 1.0]` and adjusting
+segmentation masks indices from 1-based to 0-based.
 
+Also note the renaming of the keys of the dictionary. The processed datasets are
+formatted suitably for KerasCV layers, which expect inputs in a specific dictionary
+format.
 """
-## What does one input image and corresponding segmentation mask look like?
+
+key_rename_fn = lambda inputs: {
```
[the remark this answers is not in the document]
Comment (ariG23498) [reply examples/vision/oxford_pets_image_segmentation.py:93]: What would you suggest?

[the remark this answers is not in the document]
Comment (ianstenbit) [reply examples/vision/oxford_pets_image_segmentation.py:93]: Maybe `rescale_images_and_correct_masks`?

**Comment by ariG23498** [comment]
@fchollet over to you for the final review!

**Comment by martin-gorner** [comment]
I worked with Aritra on this example. This is the final version.
LGTM

**Review by martin-gorner — APPROVED** [review]
I worked with Aritra on this example. This is the final version.
LGTM

**Comment by ariG23498** [comment]
@martin-gorner I have also added the generated files!

**Review by fchollet — APPROVED** [review]
LGTM, thank you!
